damus/Views/DMView.swift

```diff
@@ -30,49 +37,296 @@ struct DMView: View {
-                .overlay(Text(format_relative_time(event.created_at))
-                               .font(.footnote)
-                               .foregroundColor(.gray)
-                               .opacity(0.8)
-                               .offset(x: -10, y: -5), alignment: .bottomTrailing)
+                Text(format_timestamp(timestamp: event.created_at))
+                    .font(.system(size: 12))
+                    .font(.footnote)
+                    .foregroundColor(.gray)
+                
+                if !is_ours {
+                    Spacer(minLength: UIScreen.main.bounds.width * 0.1)
+                }
+            }
+        }
+    }
+    
+    func filter_content(blocks: [Block], profiles: Profiles, privkey: String?) -> (Bool, CompatibleText?) {
+        let one_note_ref = blocks
+            .filter({ $0.is_note_mention })
+            .count == 1
+        
+        var ind: Int = -1
+        var show_text: Bool = false
+        let txt: CompatibleText = blocks.reduce(CompatibleText()) { str, block in
+            ind = ind + 1
             
-            if !is_ours {
-                Spacer(minLength: UIScreen.main.bounds.width * 0.2)
+            switch block {
+            case .mention(let m):
+                if m.type == .event && one_note_ref {
+                    return str
+                }
+                if m.type == .pubkey {
+                    show_text = true
+                }
+                return str + mention_str(m, profiles: profiles)
+            case .text(let txt):
+                var trimmed = txt
+                if let prev = blocks[safe: ind-1], case .url(let u) = prev, is_image_url(u) {
+                    trimmed = " " + trim_prefix(trimmed)
+                }
+                
+                if let next = blocks[safe: ind+1] {
+                    if case .url(let u) = next, is_image_url(u) {
+                        trimmed = trim_suffix(trimmed)
+                    } else if case .mention(let m) = next, m.type == .event, one_note_ref {
+                        trimmed = trim_suffix(trimmed)
+                    }
+                }
+                if (!trimmed.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) {
+                    show_text = true
+                }
+                return str + CompatibleText(stringLiteral: trimmed)
+            case .relay(let relay):
+                show_text = true
+                return str + CompatibleText(stringLiteral: relay)
+            case .hashtag(let htag):
+                show_text = true
+                return str + hashtag_str(htag)
+            case .invoice:
+                return str
+            case .url(let url):
+                if is_image_url(url) {
+                    return str
+                } else {
+                    show_text = true
+                    return str + url_str(url)
+                }
             }
         }
+
+        return (show_text, txt)
     }
```

Comment:
this looks copy + pasted? what is this for?

Reply:
Yea it's very similar to `render_blocks` in `NoteContentView.swift` but it does something else. For example when someone sends an image with the message I want it to display the image separately from the normal text display (without the purple background). Like this (this is one note btw):

![Screenshot 2023-04-24 at 5 19 18 PM](https://user-images.githubusercontent.com/43693074/234143368-c64b884e-e259-4cf5-aa48-7c31034ee648.png)